src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb

```diff
@@ -2040,58 +2077,107 @@ def http(uri)
       http
     end
 
-    # https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-protocols-oauth-service-to-service
     def get_token(force_refresh = false)
       if @token.nil? || (Time.at(@token['expires_on'].to_i) - Time.new) <= 0 || force_refresh
         @logger.info('get_token - trying to get/refresh Azure authentication token')
-        endpoint, api_version = get_azure_authentication_endpoint_and_api_version(@azure_config)
-        uri = URI(endpoint)
-        params = {
-          'api-version' => api_version
-        }
-        uri.query = URI.encode_www_form(params)
-        @logger.debug("get_token - authentication_endpoint: #{uri}")
-
-        request = Net::HTTP::Post.new(uri.request_uri)
-        request['Content-Type'] = 'application/x-www-form-urlencoded'
-        request = merge_http_common_headers(request)
-        @logger.debug('get_token - request.header:')
-        request.each_header { |k, v| @logger.debug("\t#{k} = #{v}") }
-
-        client_id = @azure_config.client_id
-        request_body = {
-          'grant_type' => 'client_credentials',
-          'client_id'  => client_id,
-          'resource'   => get_token_resource(@azure_config),
-          'scope'      => 'user_impersonation'
-        }
-        if !@azure_config.client_secret.nil?
-          request_body['client_secret'] = @azure_config.client_secret
-        else
-          request_body['client_assertion_type'] = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer'
-          request_body['client_assertion']      = get_jwt_assertion(endpoint, client_id)
+        use_msi = @azure_config.enable_managed_service_identity
+        use_ssl = !use_msi
```

Comment:
after read this line, I guess the the msi endpoint does not support ssl?

Reply:
You are right. Because the msi endpoint is not exposed to public, it's acceptable to use "http". How do you think?